### ISSUE TITLE
Refactor dependency injection of window object in paywall hooks

### ISSUE THREAD

**Issue by cellog** [unlock-protocol/unlock#1828]
All of the browser paywall hooks use `window`, and accept it as a parameter. This can lead to extremely wonky-looking code.

There are a couple of options for solving this. The most obvious is to mock the window object, something like:

```js
beforeEach(() => {
  global.window = { ... }
})
```

I have tried this in other projects, and it occasionally led to race conditions where unrelated tests were broken by the mocks. This loser posted a question about it on Twitter: https://twitter.com/GregCello/status/1101306198538752000

and the commenter seemed very confident in this approach.

Another alternative is to pass the window global down using context, something like:

```js
import React, { createContext, useContext } from 'react'

export const WindowContext = createContext()

export default function useWindow() {
  return useContext(WindowContext)
}
```

and then in the top-level file:

```js
import { WindowContext } from '../path/to/hooks/useWindow.js'

// ...

return (
  <WindowContext.Provider value={window}>...</WindowContext.Provider>
)
```

Testing is less magical (no mocking needed), at the expense of slightly more boilerplate.  Every hook that uses it would need to add 1 line:

```js
const window = useWindow()
```

But then we have more isolation of the code. Would love to get feedback to make a decision on this one, @julien51 @akeem and @cnasc

**Comment by julien51** [comment]
Thanks for the research! The `global` mocking approach seems fine to me, and resonates with what I've seen done in the past... 

My worry about the context approach is that it means that our components can only work then if they are wrapped inside the provider?

**Comment by cellog** [comment]
Correct, the components would only work when wrapped, however, the errors would be pretty immediate, and would cause integration tests to fail immediately once we have them. The provider would be put into _document.js along with the others.

**Comment by cellog** [comment]
after further testing, I can confirm that it is not possible to override `global.window`

```js
global.window = 'hi'
console.log(global.window)
```

outputs the window object. So, perhaps I can modify the fields of window, I will try that next

**Comment by cellog** [comment]
that also does not work. window is frozen, apparently

**Comment by cellog** [comment]
I think it's still possible to mock, going to try something like https://gist.github.com/remarkablemark/5cb571a13a6635ab89cf2bb47dc004a3 or possibly using node environment in this test only

**Comment by cellog** [comment]
update: mocking methods is working, but I can't figure out how to make `location` and `parent` changeable yet

**Comment by cellog** [comment]
ok, I've spent 2 hours trying to get mocking of window to work, and here is what I've discovered:

1. It's possible to mock out methods on the window object
2. jest itself uses window methods addEventListener and removeEventListener, so this makes them unassertable (they are called a variable number of times, depending on the number of tests in the file)
3. even though the descriptors for `window.parent` and `window.location` say they are both writable and configurable, they still do not respond to `window.parent = {}`, which is a characteristic of non-writable values.

It doesn't seem like this is a viable path, I recommend we handle this by going down the context route, and ensuring our components work by creating a simple test that pulls in the top-level component, reads the window config and asserts that we actually distribute the `window` object from it.

How does that sound?
